web-app/HeimScripts/heatmap/run.R

```diff
@@ -1,187 +1,254 @@
+}
+
+discardNodeAndSubject <- function(label) {
+  label <- strsplit(label,"_")
+  endOfSubject <-
+    length(label) - 2  #last too elements are node and subset.
+  label <- label[1:endOfSubject]
+  paste(label, collapse = "_")
 }
 
-buildFields <- function(df){
-  df <- melt(df, na.rm = T, id=c("Row.Label","Bio.marker","SIGNIFICANCE","MEAN","SD")) # melt implicitly casts characters to factors to make your like more exciting, in order to encourage more adventures it does not have characters.as.factors=F param.
-  ZSCORE <- (df$value - df$MEAN)/df$SD
+buildFields <- function(df) {
+  df <- melt(
+    df, na.rm = T, id = c("Row.Label",
+                          "Bio.marker",
+                          "SIGNIFICANCE",
+                          "MEAN",
+                          "SD")
+  )  # melt implicitly casts
+     # characters to factors to make your like more exciting,
+     # in order to encourage more adventures it does not
+     # have characters.as.factors=F param.
+  ZSCORE <- (df$value - df$MEAN) / df$SD
   df["MEAN"] <- NULL
-  df["SD"]   <-NULL
-  names(df) <- c("PROBE","GENESYMBOL","SIGNIFICANCE","PATIENTID","VALUE")
+  df["SD"]   <- NULL
+  names(df) <-
+    c("PROBE","GENESYMBOL","SIGNIFICANCE","PATIENTID","VALUE")
   df["ZSCORE"] <- ZSCORE
   return(df)
 }
 
-writeDataForZip <- function(df, patientIDs) {
-  pidCols    <- as.character(patientIDs)
-  t          <- df
-  t[pidCols] <- lapply(t[pidCols], function(v) { (v - df$MEAN) / df$SD })
-  allCols    <- c(c("Row.Label", "Bio.marker", "MEAN", "SD", "SIGNIFICANCE"), pidCols)
-  write.table(t[allCols], "heatmap_data.tsv", sep="\t", na="", row.names=F, col.names=T)
-
-  allCols <- c(c("Row.Label"), pidCols)
-  write.table(df[allCols], "heatmap_orig_values.tsv", sep="\t", na="", row.names=F, col.names=T)
+writeDataForZip <- function(df, zScores, patientIDs) {
+  pidCols <- as.character(patientIDs)
+  df      <- df[ , -which(names(df) %in% pidCols)]  # Drop patient columns
+  df      <- cbind(df,zScores)                      # Replace with zScores
+  write.table(
+    df,
+    "heatmap_data.tsv",
+    sep = "\t",
+    na = "",
+    row.names = FALSE,
+    col.names = TRUE
+  )
 }
 
 writeRunParams <- function(max_rows, sorting) {
-  params <- list(
-    max_rows = max_rows,
-    sorting  = sorting
-  )
-
+  params <- list(max_rows = max_rows,
+                 sorting  = sorting)
   if (exists("preprocessed") && exists("preprocessing_params")) {
     params <- c(params, preprocessing_params)
   }
-
-  write(toJSON(params, pretty=TRUE), 'params.json')
+  write(toJSON(params, pretty = TRUE), 'params.json')
 }
 
 fixString <- function(str) {
-  str <- gsub("[^a-zA-Z0-9-]", "", str, perl=TRUE)
-  return(str)
+  gsub("[^a-zA-Z0-9-]", "", str, perl = TRUE)
 }
 
-buildExtraFields <- function(df){
-  FEATURE <- rep("Cohort", nrow(df) )
+buildExtraFields <- function(df) {
+  FEATURE <- rep("Cohort", nrow(df))
   PATIENTID <- as.character(df$PATIENTID)
```

Comment:
as.character? what is the type of PATIENTID column, if it's not character? If it is factor: maybe convert it to character already in buildFields (where this column is created)? Then the class conversion only needs to take place once  + it has the advantage that no weird behavior caused by factors will occur when another operation is added that uses the patient IDs